### PR TITLE
ensure imbalances are properly accounted for

### LINKED ISSUE #2182: Is `TotalIssuance` updated correctly? after #1979 change, the `Imbalance` trait would call `drop` in an unexpect way
in issue #1979 , it proposal the `TotalIssuance` not change currectly, so there is a pr #2048 to fix this.
we can see that in #2048 provide the `drop` trait to update `TotalIssuance`:
```rust
impl<T: Subtrait<I>, I: Instance> Drop for PositiveImbalance<T, I> {
	/// Basic drop handler will just square up the total issuance.
	fn drop(&mut self) {
		<TotalIssuance<ElevatedTrait<T, I>, I>>::mutate(|v| *v = v.saturating_add(self.0));
	}
}

 impl<T: Subtrait<I>, I: Instance> Drop for NegativeImbalance<T, I> {
	/// Basic drop handler will just square up the total issuance.
	fn drop(&mut self) {
		<TotalIssuance<ElevatedTrait<T, I>, I>>::mutate(|v| *v = v.saturating_sub(self.0));
	}
}	}
```
but the `Imbalance` trait like:
```rust
#[must_use]
pub trait Imbalance<Balance>: Sized {
	/// The oppositely imbalanced type. They come in pairs.
	type Opposite: Imbalance<Balance>;

	fn drop_zero(self) -> Result<(), Self>;
	/// Consume `self` and return two independent instances; the first
	/// is guaranteed to be at most `amount` and the second will be the remainder.
	fn split(self, amount: Balance) -> (Self, Self);  // this one
	fn merge(self, other: Self) -> Self;  // this one
	fn maybe_merge(self, other: Option<Self>) -> Self {  // this one
		// ...
	}
	fn subsume(&mut self, other: Self);
	fn offset(self, other: Self::Opposite) -> Result<Self, Self::Opposite>;  // and this one
	fn peek(&self) -> Balance;
}
```
in this trait, we should notice that some function use `self` like `split`, `merge`, `offset` as first param to move/consume the object, and **create** new object to return:
```rust
// impl for PositiveImbalance
impl<T: Trait<I>, I: Instance> Imbalance<T::Balance> for PositiveImbalance<T, I> {
// ...
// drop_zero is not important
// notice this:
	fn split(self, amount: T::Balance) -> (Self, Self) {
		let first = self.0.min(amount);
		let second = self.0 - first;
		(Self(first), Self(second))  // create new object, and the old obj `self` would drop after this function!
	}
// and other func is in the same way
}
```
In fact, the function which is start as `self` as the first param, this object would be dropped! so that if a `imbalance` object call `merge`, `split` to combine or split objects, to call more times `drop` to modify the `TotalIssuance`

for example, if I call some func and get an imbalance object:
```rust
// for example `Currency::deposit_creating()`, would return PositiveImbalance, in a closure
{
    let p: PositiveImbalance = Currency::deposit_creating(...); // PositiveImbalance contains 10
    // then I call split
    let (a, b) = p.split(One::one);
}
// then, after this closure
// obj p, a, b would all be dropped!
// then the `drop` would call three times!,  drop for p(TotalIssuance increase 10), drop for a(TotalIssuance increase 1) and drop for b(TotalIssuance increase 9), and totally increase 20 for `TotalIssuance` 
```
and an other example in srml/contract/a.ccount_db.rs L120
```rust
	fn commit(&mut self, s: ChangeSet<T>) {
		let mut total_imbalance = SignedImbalance::zero();
		for (address, changed) in s.into_iter() {
			let trieid = <Self as AccountDb<T>>::get_or_create_trieid(&self, &address);
			if let Some(balance) = changed.balance {
				let (imbalance, outcome) = T::Currency::make_free_balance_be(&address, balance);
				total_imbalance = total_imbalance.merge(imbalance); // this merge
```
in the loop, every time would drop the object `total_imbalance`, and `imbalance` inner the function merge, cause unexpect modify for `TotalIssuance`

Above all, the `drop` trait should not use with function which the ownership can be moved.
I think should change the `Imbalance` trait, not allow `self` for first param

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #2182

### CODE REVIEW

**Comment by rphmeier** [comment]
(adding it the milestone just to make sure it gets looked at prior to release)

File: node/runtime/src/lib.rs [hunk 1]
```diff
@@ -59,8 +59,8 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("node"),
 	impl_name: create_runtime_str!("substrate-node"),
 	authoring_version: 10,
-	spec_version: 56,
-	impl_version: 59,
+	spec_version: 57,
+	impl_version: 60,
```
Comment (gavofyork) [inline node/runtime/src/lib.rs]:
```suggestion
	impl_version: 0,
```

Can reset when spec version changes.

File: node/runtime/src/lib.rs [hunk 2]
```diff
@@ -59,8 +59,8 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("node"),
 	impl_name: create_runtime_str!("substrate-node"),
 	authoring_version: 10,
-	spec_version: 56,
-	impl_version: 59,
+	spec_version: 57,
```
Comment (gavofyork) [inline node/runtime/src/lib.rs]:
```suggestion
	spec_version: 56,
```

**Comment by gavofyork** [comment]
just re-read this code and saw `mem::forget` is used now, so yeah the `spec_version` should have been bumped. shouldn't be a big issue to leave it this once as there are no deployed networks with the previous version.

**Comment by rphmeier** [comment]
I did bump the spec version in `node-runtime` -- where else?

**Pull request merged**